persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/DiscreteCopyNumberSparkRepository.java

```diff
@@ -0,0 +1,207 @@
+            .collectAsList().stream().map(r -> r.getString(0)).collect(Collectors.toList());
+        String studyId = studyIds.get(0);
+        // filter by alterations
+        Dataset<Row> cnaData = spark.read().parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + ParquetConstants.DATA_CNA);
+        if (!CollectionUtils.isEmpty(alterations)) {
+            cnaData = cnaData.where(cnaData.col("VALUE").isin(alterations.toArray()));
+        }
+        // filter by patientIds
+        Dataset<Row> clinicalSample = spark.read().parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + ParquetConstants.CLINICAL_SAMPLE);
+        if (!CollectionUtils.isEmpty(patientIds)) {
+            clinicalSample = clinicalSample
+                .where(clinicalSample.col("PATIENT_ID").isin(patientIds.toArray()));
+        }
+        casels = casels.join(clinicalSample, "SAMPLE_ID");
+        cnaData = cnaData.join(casels, "SAMPLE_ID");
+
+        String mainCol = Arrays.asList(cnaData.columns()).contains("Entrez_Gene_Id") ? "Entrez_Gene_Id" : "Hugo_Symbol";
+        cnaData.createOrReplaceTempView("cna");
+        StringBuilder sb = new StringBuilder();
+        sb.append("SELECT " + mainCol + ", VALUE, COUNT(DISTINCT(PATIENT_ID)) as numberOfAlteredCases ");
+        sb.append("FROM cna ");
+        sb.append("GROUP BY " + mainCol + ", VALUE ");
+        sb.append("SORT BY " + mainCol);
+
+        Dataset<Row> result = spark.sql(sb.toString());
+
+        List<CopyNumberCountByGene> copyNumberCountByGenes = result.collectAsList().stream()
+            .map(r -> mapToCopyNumberCountByGene(r, mainCol)).collect(Collectors.toList());
+
+        return getGeneDetails(copyNumberCountByGenes, mainCol, entrezGeneIds);
+    }
+    
+    private List<CopyNumberCountByGene> getGeneDetails(List<CopyNumberCountByGene> copyNumberCountByGenes, String colName, List<Integer> entrezGeneIds) {
+        if ("Hugo_Symbol".equalsIgnoreCase(colName)) {
+            Set<String> hugoSymbols = copyNumberCountByGenes.stream()
+                .map(CopyNumberCountByGene::getHugoGeneSymbol).collect(Collectors.toSet());
+            List<Gene> genes = geneRepository.fetchGenesByHugoGeneSymbols(new ArrayList<>(hugoSymbols), PersistenceConstants.SUMMARY_PROJECTION);
+            Map<String, Gene> geneMap = genes.stream()
+                .collect(Collectors.toMap(Gene::getHugoGeneSymbol, Function.identity()));
+            
+            List<CopyNumberCountByGene> tobeRemoved = new ArrayList<>();
+            for (CopyNumberCountByGene cncg : copyNumberCountByGenes) {
+                Gene gene = geneMap.get(cncg.getHugoGeneSymbol());
+                if (gene != null) {
+                    cncg.setCytoband(gene.getCytoband());
+                    cncg.setEntrezGeneId(gene.getEntrezGeneId());
+                } else {
+                    tobeRemoved.add(cncg);
+                }
+            }
+            copyNumberCountByGenes.removeAll(tobeRemoved);
```

Comment:
why not add genes filter directly to spark sql query?

Reply:
this was added to getGeneDetails as entrezGeneIds weren't included in all data files.